KOGITO-7910: Patch PR for Kubernetes service discovery guide with doc improvements (#194)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/kubernetes-service-discovery.adoc

= Kubernetes service discovery in {context}
:compat-mode!:
// Metadata:
:description: Explain what is and how the service discovery works
:keywords: kogito, workflow, quarkus, serverless, service-discovery, enhanced-service-discovery
// links
:quarkus_issue_url: https://github.com/quarkusio/quarkus/issues/27457

The Kubernetes service discovery allows you to have a static URI, defining a Kubernetes resource, which is used to perform HTTP requests. The Kubernetes resource defined in the URI is queried in the current Kubernetes cluster and translated in a valid URL. 

The Kubernetes service discovery feature works during the workflow application startup, in which this feature scans all the Quarkus configuration in search of the following URI pattern:

> .URI pattern in Kubernetes service discovery

Reviewer: ```suggestion
.KubeURI Pattern
```

Reviewer 2: Please don't use `KubeURI`, this means nothing.

Author: So should we keep `.URI pattern in Kubernetes service discovery`?

                                                                   - labels=label-name=label-value;other-label=other-value
----

> Therefore, you must keep in mind that if application startup time matters, then consider to use a static URL instead. 

Reviewer: isn't this sentence lost here?

Author: Done, changed the paragraph above the URI pattern as follows:

The Kubernetes service discovery feature works during the workflow application startup, in which this feature scans all the Quarkus configuration in search of the URI pattern. Therefore, you must keep in mind that if application startup time matters, then consider to use a static URL instead.

Following is the URI pattern in Kubernetes service discovery:

The following scheme values are supported in the URI pattern:

* `kubernetes`
* `openshift`
* `knative`

The following resources are supported for the Kubernetes GVK (Group, Version, and Kind):

* `v1/service`
* `serving.knative.dev/v1/service`
* `v1/pod`
* `apps/v1/deployment`
* `apps.openshift.io/v1/deploymentconfig`
* `apps/v1/statefulset`
* `route.openshift.io/v1/route`
* `networking.k8s.io/v1/ingress`

[IMPORTANT]
====
The Kubernetes GVK is mandatory and cannot be empty. In case the Kubernetes GVK is empty, the engine throws a validation issue and discovery is not performed. The last value in the URI is always considered as resource name. 

The `namespace` in the URI is optional, however, if `namespace` contains an empty value, the current namespace or context is used.
====

> Query parameters in URI::

Reviewer: ```suggestion
KubeURI Query Parameters::
```

Reviewer 2: Same thing. This is just a URI that represents a Kubernetes object in the cluster.

Author: So, should we keep as-is?

+

> --

Reviewer: better to keep or add a small tittle here, appending information for different contexts may lead to some confusion.

Author: Removing other information from this section.

> The query parameters are also known as query string. The query parameters are defined the similar way with URLs to assign value to specific attributes.

Author: Do we need to capitalize Query String?

> The query parameters are also known as query string. The query parameters are defined the similar way with URLs to assign value to specific attributes.

Reviewer 2: I don't think so. Indeed in the RFC is just `query`. https://www.rfc-editor.org/rfc/rfc2616#section-3.2.2

But I think `query string` is just fine.

> The query parameters are also known as query string. The query parameters are defined the similar way with URLs to assign value to specific attributes.

Reviewer: > Do we need to capitalize Query String?

no.

> The query parameters are also known as query string. The query parameters are defined the similar way with URLs to assign value to specific attributes.

Author: Changed it to "Also known as query string`.

* *Custom labels*: The custom labels are used to filter services in case there are more than one service with the same label selector but exposing different ports. In this case, you can instruct the engine that if more than one service is found, then the engine must use the service containing the provided label. 
+
The label is defined with the following expression and in case of multiple labels, you can use semicolon (;):
+
`labels=label-name=namevalue;another-label=another-value`
+
.Example label definition in URI
[source,shell]
----
kubernetes:v1/pod/<namespace>/<pod-name>?labels=label-name=test-label
----
+
Using the previous URI example, if there are more than one service exposing the given pod, the `label-name=test-label` label is used to filter the service. If the label does not exist, the first found service is used.

* *Custom port name*: The custom port name is used to determine which port to use when multiple ports are configured in the target service or container. You can configure the port name to be queried using the following pattern:
+
`port-name=<PORT_NAME>`
--

> Based on the resource to be discovered, the Kubernetes service discovery follows specific paths as shown in the following figure:

Author: Done, changed it to a point.

image::cloud/sw-discovery-flow.jpg[]

> There is no specific configuration required for the Kubernetes service discovery except by using the expected URI pattern. However, the `okhttp` communication interceptor, which logs the communication between the application and the Kubernetes API is disabled by default.

Reviewer: also, these following sentences are a little bit disconnected without a tittle, maybe keep the Configuration subtitle and group it and the dependency exclusion?

Author: Added a section for configuration.

> You can enable the `okhttp` communication interceptor if there is a need to debug it by setting the following application property:

Reviewer: ```suggestion
You can enable the `okhttp` communication interceptor if there is a need to debug communication betwwen the client and the Kubernetes API by setting the following application property:
```

Author: Done

> .Property to enable `okhttp` communication interceptor

Reviewer: ```suggestion
.Application Property to enable `okhttp` communication interceptor
```

Author: Done

[source,shell]
----
quarkus.log.category."okhttp3.OkHttpClient".level=INFO
----

[[ref-example-kubernetes-service-discovery]]

> == Example of Kubernetes service discovery in {context}

Reviewer: which example are you referring to?

Author: referring to the `serverless-workflow-greeting-quarkus` example, which is also used in the URI in the section.

Reviewer: not sure if I understand it.

Author: We have a URI in this section as follows which contains `serverless-workflow-greeting-quarkus`, therefore added example:

![image](https://user-images.githubusercontent.com/59050394/190379037-96714827-b0ce-4aac-a156-530d0bc17233.png)

> The Kubernetes service discovery is performed at the _STATIC_INIT_ time of Quarkus during the workflow application startup. First, the service discovery scans the Quarkus configuration values and searches for the Kubernetes URI pattern. If the URI pattern is found, the engine parses the URI, queries the Kubernetes API for the given resource, and overrides the given property.

Reviewer: ```suggestion
The Kubernetes service discovery is performed at the _STATIC_INIT_ time of Quarkus during the workflow application startup. First, the service discovery scans the Quarkus configuration values and searches for the Kubernetes URI pattern. If the URI pattern is found, the engine parses the URI, queries the Kubernetes API searching for the given resource, and overrides the given property.
```

Author: Done

> The Kubernetes service discovery is performed at the _STATIC_INIT_ time of Quarkus during the workflow application startup. First, the service discovery scans the Quarkus configuration values and searches for the Kubernetes URI pattern. If the URI pattern is found, the engine parses the URI, queries the Kubernetes API for the given resource, and overrides the given property.

Reviewer: ```suggestion
The Kubernetes service discovery is performed at the _STATIC_INIT_ time of Quarkus during the workflow application startup. First, the service discovery scans the Quarkus configuration values and searches for the Kubernetes URI pattern. If the URI pattern is found, the engine parses the URI, queries the Kubernetes API for the given resource, and overrides the given application property.
```

Author: Done

[NOTE]
====
Note that the URI was translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io` when the application started, and will be used at runtime any time it needs it.
In the previous example, the URI is translated to `http://greeting-quarkus-cli.serverless-workflow-greeting-quarkus.10.99.154.147.sslip.io` when the application started. The translated URI is used at runtime, when needed.
====

> == Discovery Precedence

Reviewer: I would keep it.

[tabs]
====

> Kubernetes service discovery is enabled::

Reviewer: isn't better to remove the "is"?

Author: I think the sentence looks better with is.

+
[source,shell]
----
(main) sw-service-discovery 1.0.0-SNAPSHOT on JVM (powered by Quarkus 999-SNAPSHOT) started in 2.360s. Listening on: http://0.0.0.0:8080
----

> Kubernetes service discovery is disabled::

Reviewer: same

Author: I think the sentence looks better with is.
